$\large{\textbf{Answer 1:}}$

In [ ]:
!pip install -q pyomo

     |████████████████████████████████| 9.6 MB 4.9 MB/s 
     |████████████████████████████████| 49 kB 3.0 MB/s 


In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
model = ConcreteModel()

In [ ]:
N = 3 #no of variables
M = 3 #no of constraints

In [ ]:
lb = np.array([2,1,1])
ub = np.array([np.inf, np.inf, np.inf])

In [ ]:
obj_coef_c = np.array([2, -2, 2])

In [ ]:
cons_coef_A = np.array([[2, -1, 2], [1, -4, 2], [-2, 2, 1]])
coef_rhs_b = np.array([6, -15, -8])

In [ ]:
col_indices = np.arange(N)
row_indices = np.arange(M)

In [ ]:
model.x = Var(col_indices)

In [ ]:
model.constraints = ConstraintList()
for i in row_indices:
  model.constraints.add(sum(cons_coef_A[i][j]*model.x[j] for j in col_indices) <= coef_rhs_b[i])

In [ ]:
model.obj = Objective(expr = summation(obj_coef_c,model.x), sense=minimize)

In [ ]:
for i in col_indices:
  model.x[i].setub(ub[i])
  model.x[i].setlb(lb[i])

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {1, 2, 3}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     2 :  None :  None : False :  True :  Reals
          1 :     1 :  None :  None : False :  True :  Reals
          2 :     1 :  None :  None : False :  True :  Reals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 2*x[0] - 2*x[1] + 2*x[2]

1 Constraint Declarations
    constraints : Size=3, Index=constraints_index, Active=True
        Key : Lower : Body                    : Upper : Active
          1 :  -Inf :  2*x[0] - x[1] + 2*x[2] :   6.0 :   

In [ ]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155676 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [ ]:
opt = SolverFactory('glpk', executable='/usr/bin/glpsol')
result = opt.solve(model)

print(result)
print("Solver Status:",result.solver.status)
print('Termination Condition:', result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 10
  Sense: minimize
Solver: 
- Status: ok
  Termination condition: other
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.05112504959106445

Solver Status: ok
Termination Condition: other


$\text{The solver status and termination condition are not meaningful in this case as we are not getting a solution but still the solver status is coming out to be OK.}\\ \text{The termination condition is showing OTHER which does not clear out the reason but we can only get the possibilities that either the problem is infeasible or unbounded.}$

In [ ]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 155879 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [ ]:
cbcSolver = SolverFactory('cbc')

In [ ]:
result2 = cbcSolver.solve(model)

print(result2)
print('Solver Status: ',  result2.solver.status)
print('Solver termination condition: ', result2.solver.termination_condition)

  - termination condition: infeasible
  - message from solver: <undefined>



Problem: 
- Name: unknown
  Lower bound: None
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 3
  Sense: minimize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.0590815544128418

Solver Status:  warning
Solver termination condition:  infeasible


$\text{Here the messages displayed are meaningful. The solver status is displaying WARNING which may be because of the termination condition that is coming out to be INFEASIBLE.} \\ \text{The possible reason for the system to be infeasible is that there is no solution which satisfies all the constraints simultaneously.}$

$\large{\textbf{Answer 2.a}} \\  \text{Changimg the sense of objective function to maximum.}$

In [ ]:
model.obj.set_sense(maximize)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {1, 2, 3}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     2 :  None :  None : False :  True :  Reals
          1 :     1 :  None :  None : False :  True :  Reals
          2 :     1 :  None :  None : False :  True :  Reals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 2*x[0] - 2*x[1] + 2*x[2]

1 Constraint Declarations
    constraints : Size=3, Index=constraints_index, Active=True
        Key : Lower : Body                    : Upper : Active
          1 :  -Inf :  2*x[0] - x[1] + 2*x[2] :   6.0 :   

In [ ]:
result = cbcSolver.solve(model)

print(result)
print('Solver status: ', result.solver.status)
print('Solver termination condition: ', result.solver.termination_condition)

  - termination condition: infeasible
  - message from solver: <undefined>



Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: None
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 3
  Sense: maximize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.028020381927490234

Solver status:  warning
Solver termination condition:  infeasible


$\text{Here the messages displayed are meaningful. The solver status is displaying WARNING which may be because of the termination condition that is coming out to be INFEASIBLE.} \\ \text{It is obvious that the problem will still be infeasible as the sense of the objective fuction has nothing to do with the infeasibility of the problem.} \\ \text{Had it been unbounded then changing the sense might have helped in solving the problem.}$

$\large{\textbf{Answer 2.b :}} \\ \text{Setting the upper bounds of } x_2 \text{ as } 5. \\ \text{Setting the upper bounds of } x_3 \text{ as } 10.$ 

In [ ]:
model.obj.set_sense(minimize)
model.x[1].setub(5)
model.x[2].setub(10)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {1, 2, 3}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     2 :  None :  None : False :  True :  Reals
          1 :     1 :  None :     5 : False :  True :  Reals
          2 :     1 :  None :    10 : False :  True :  Reals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 2*x[0] - 2*x[1] + 2*x[2]

1 Constraint Declarations
    constraints : Size=3, Index=constraints_index, Active=True
        Key : Lower : Body                    : Upper : Active
          1 :  -Inf :  2*x[0] - x[1] + 2*x[2] :   6.0 :   

In [ ]:
result = cbcSolver.solve(model)

print(result)
print('Solver status: ', result.solver.status)
print('Solver termination condition: ', result.solver.termination_condition)

  - termination condition: infeasible
  - message from solver: <undefined>



Problem: 
- Name: unknown
  Lower bound: None
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 3
  Sense: minimize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.02541375160217285

Solver status:  warning
Solver termination condition:  infeasible


$\text{Here the messages displayed are meaningful. The solver status is displaying WARNING which may be because of the termination condition that is coming out to be INFEASIBLE.} \\ \text{Here also the problem remains to be infeasible.} \\ \text{This is because we have only restricted the feasible region and when there is no solution in the bigger region there is no chance that the solution will be in a smaller one.} \\ \text{Instead of adding upper bounds extending the lower bounds might have helped.}$

$\large{\textbf{Answer 2.c :}} \\ \text{Adding a new constraint: } \\ -x_2 + x_3 \leq 16$

In [ ]:
model.x[1].setub(np.inf)
model.x[2].setub(np.inf)

In [ ]:
new_cons_coef = np.array([0,-1,1])
new_cons_rhs = 16
model.constraints.add(summation(new_cons_coef,model.x) <= new_cons_rhs)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     2 :  None :  None : False :  True :  Reals
          1 :     1 :  None :  None : False :  True :  Reals
          2 :     1 :  None :  None : False :  True :  Reals

1 Objective Declarations
    obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 2*x[0] - 2*x[1] + 2*x[2]

1 Constraint Declarations
    constraints : Size=4, Index=constraints_index, Active=True
        Key : Lower : Body                    : Upper : Active
          1 :  -Inf :  2*x[0] - x[1] + 2*x[2] :   6.0 :

In [ ]:
result = cbcSolver.solve(model)

print(result)
print('Solver status: ', result.solver.status)
print('Solver termination condition: ', result.solver.termination_condition)

  - termination condition: infeasible
  - message from solver: <undefined>



Problem: 
- Name: unknown
  Lower bound: None
  Upper bound: 11.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 4
  Number of nonzeros: 0
  Sense: minimize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.027406692504882812

Solver status:  warning
Solver termination condition:  infeasible


$\text{Here the messages displayed are meaningful. The solver status is displaying WARNING which may be because of the termination condition that is coming out to be INFEASIBLE.} \\ \text{Here also the problem remains to be infeasible.} \\ \text{This is because we have only restricted the feasible region and when there is no solution in the bigger region there is no chance that the solution will be in a smaller one.} \\ \text{Instead of adding a constraint, deactivating/relaxing a constraint might have helped.}$

$\large{\textbf{Answer 2.d :}} \\ \text{Deactivating the objective function and adding a new objective function and also deactivating constraint 1 and adding a new constraint.} \\ \text{New objective function: min } x_1 + 2x_2 + 3x_3 \\ \text{New constraint: } x_2 + x_3 \geq 25$

In [ ]:
model.constraints[4].deactivate()

new_obj_coef = np.array([1,2,3])
next_cons_coef = np.array([0,1,1])
next_cons_rhs = 25

In [ ]:
model.new_obj = Objective(expr = summation(new_obj_coef,model.x), sense = minimize)
model.constraints.add(summation(next_cons_coef,model.x) >= next_cons_rhs)

model.obj.deactivate()
model.constraints[1].deactivate()

model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     2 :  None :  None : False :  True :  Reals
          1 :     1 :  None :  None : False :  True :  Reals
          2 :     1 :  None :  None : False :  True :  Reals

2 Objective Declarations
    new_obj : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : x[0] + 2*x[1] + 3*x[2]
    obj : Size=1, Index=None, Active=False
        Key  : Active : Sense    : Expression
        None :  False : minimize : 2*x[0] - 2*x[1] + 2*x[2]

1 Constraint Declarations
    constraints : Size=5, I

In [ ]:
result = cbcSolver.solve(model)

print(result)
print('Solver status: ', result.solver.status)
print('Solver termination condition: ', result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: 79.5
  Upper bound: 79.5
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 3
  Sense: minimize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 2
  Error rc: 0
  Time: 0.020871639251708984
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

Solver status:  ok
Solver termination condition:  optimal


$\text{Here the messages displayed are meaningful. The solver status is displaying OK and the termination condition is coming out to be OPTIMAL.} \\ \text{The possible reason that the problem is solved now might be the the first constraint.} \\ \text{It may be the reason that the problem did not has a solution.}$

In [ ]:
print('\nObjective = ', model.new_obj())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)
print('\nConstraints')
model.constraints.display()


Objective =  79.5

Decision Variables
x[ 0 ] =  28.5
x[ 1 ] =  24.0
x[ 2 ] =  1.0

Constraints
constraints : Size=5
    Key : Lower : Body  : Upper
      2 :  None : -65.5 : -15.0
      3 :  None :  -8.0 :  -8.0
      5 :  25.0 :  25.0 :  None


$\textbf{Objective function value} = 79.5 \\ \textbf{Value of decision variables:} \\ x_1 = 28.5 \\ x_2 = 24 \\ x_3 = 1 \\ \textbf{Activity of constraints:} \\ \textbf{Constraint 2: } \text{It is inactive as its value is -65.5 which is strictly less than the RHS}= -15. \\ \textbf{Constraint 3: } \text{It is active as its value is -8 which is exactly equal to the RHS.} \\ \textbf{Constraint 5: } \text{It is active as its value is -25 which is exactly equal to the RHS.}$